clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -268,34 +283,42 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
     /**
      * Get a list of batches which have been sitting in the accumulator too long and need to be expired.
      */
-    public List<ProducerBatch> expiredBatches(int requestTimeout, long now) {
+    public List<ProducerBatch> expiredBatches(long now) {
         List<ProducerBatch> expiredBatches = new ArrayList<>();
+
+        // Expiration of inflight batches happens here in the order of createdMs if the final state of the
+        // batch is not known by (batch's creation time + deliveryTimeoutMs).
+        while (!soonToExpireInFlightBatches.isEmpty() &&
```

Comment:
To be honest, this lazy expiration seems like overkill. It should be a rare case where we actually have entries in `soonToExpireInFlightBatches` because of the other optimization to only add to it when the delivery timeout will expire prior to the request timeout. And if the producer is in a situation where batches are being expired, then the performance of removal for a particular batch is probably not a major concern. Maybe some benchmarking would show whether it is a worthwhile optimization.

Reply:
@hachikuji I agree in most case we probably do not need this. It is probably only useful for large producers (e.g. mirror maker which has thousands of partitions to send). There may still be value to have this:

1. Since delivery.timeout.ms and request.timeout.ms are both user configurations. It would be good to guard against some configurations. (e.g. request.timeout.ms=delivery.timeout.ms, linger.ms=0). 

2. It seems that in some scenarios this would help. For example, when the brokers are being rolling bounced, there will be a lot of retried batches. Those batches may have `remainingDeliveryTimeoutMs` < `request.timeout.ms`. We may have even more than one batches per partition to insert if max.in.flight.request is greater than 1.

A benchmark would be useful. But in general I think it is safer to have this optimization given it does not increase too much complexity.